storage/badger/monotonic_consumer_progress.go

```diff
@@ -0,0 +1,54 @@
+package badger
+
+import (
+	"fmt"
+
+	"github.com/dgraph-io/badger/v2"
+
+	"github.com/onflow/flow-go/module/counters"
+)
+
+// MonotonicConsumerProgress represents the consumer progress with strict monotonic counter.
+type MonotonicConsumerProgress struct {
+	consumerProgress *ConsumerProgress
+
+	// used to skip heights that are lower than the current height
+	counter counters.StrictMonotonousCounter
+}
+
+// NewMonotonicConsumerProgress creates a new MonotonicConsumerProgress which inserts the default
+// processed index to the storage layer and creates new counter with defaultIndex value.
+//
```

Comment:
can you add a warning here that the consumer progress and associated db entry must not be accessed outside of calls to the returned object, otherwise the state may become inconsistent.

Reply:
[8ad035c](https://github.com/onflow/flow-go/pull/5752/commits/8ad035cb2ee7386b451cc28fd1740487aef30217)